spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/ExtendedScanRelation.scala

```diff
@@ -0,0 +1,29 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.spark.sql.catalyst.plans.logical
+
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2ScanRelation
+
+case class ExtendedScanRelation(relation: DataSourceV2ScanRelation) extends LogicalPlan {
```

Comment:
I am a bit worried about adding this class as I am not sure we want to maintain it in Spark later. There is another idea how to solve the rewrite rule: we can simply disable column pruning for `DynamicFileFilter` nodes. I think it should be sufficient to extend the node references to also cover all output attributes of the scan.

```
AttributeSet(scanRelation.output ++ fileFilterPlan.output)
```

I've submitted a PR with this idea to your branch, @rdblue. Feel free to discard/modify as needed.

Reply:
I think the more general problem is that Spark couldn't rewrite `DynamicFileFilter` node at all, which led to some situations where the plan worked fine and some where it would fail. I think we should generally fit the pattern of being able to rewrite nodes.

Reply:
I get the intention here now. I did a quick look and see that Spark injects the extended planner strategies before the data source V2 strategy. That got me thinking that we can get rid of this extra logical plan and just use the scan node directly and apply the new logic only if we have a scan that implements `SupportsFileFilter`.

```
case PhysicalOperation(project, filters, DataSourceV2ScanRelation(_, scan: SupportsFileFilter, output)) =>
  // projection and filters were already pushed down in the optimizer.
  // this uses PhysicalOperation to get the projection and ensure that if the batch scan does
  // not support columnar, a projection is added to convert the rows to UnsafeRow.
  val batchExec = ExtendedBatchScanExec(output, scan)
  withProjectAndFilter(project, filters, batchExec, !batchExec.supportsColumnar) :: Nil

  ....

private def withProjectAndFilter(
    project: Seq[NamedExpression],
    filters: Seq[Expression],
    scan: LeafExecNode,
    needsUnsafeConversion: Boolean): SparkPlan = {
  val filterCondition = filters.reduceLeftOption(And)
  val withFilter = filterCondition.map(FilterExec(_, scan)).getOrElse(scan)

  if (withFilter.output != project || needsUnsafeConversion) {
    ProjectExec(project, withFilter)
  } else {
    withFilter
  }
}
```

This also takes care of adding a projection if needed.

Reply:
I that's close how a native Spark solution would work.

Reply:
There are no issues we use the same scan node in multiple places given the new logic and allowing rewrites, right?

Reply:
Scan nodes shouldn't be reused in general, but it is usually okay if there is not a join. The problem is reused attribute IDs.

Reply:
+1 to using this rule that gets rid of the extra node.